libass/ass.c

```diff
@@ -1535,8 +1535,9 @@ static ASS_Track *parse_memory(ASS_Library *library, char *buf)
  * \return newly allocated track
 */
 ASS_Track *ass_read_memory(ASS_Library *library, char *buf,
-                           size_t bufsize, char *codepage)
+                           size_t bufsize, const char *codepage)
 {
+    // Documentation promises buf isn't modified other than being passed to iconv
```

Comment:
I’d prefer to declare `const char *buf`. The `iconv` call can stay an implementation detail, and the caller shouldn’t need to care.

It may be worth mentioning that older versions of the spec [did in fact include `const`](https://pubs.opengroup.org/onlinepubs/007908799/xsh/iconv.html) in the `iconv` prototype but this was [changed after complaints](http://www.opengroup.org/infosrv/XPG4_Brand/resolutions/XoTGBase94-001.ps.Z) that this was inconsistent with other text in the spec that mentioned “The type of _inbuf_ and _outbuf_, **char \*\***” which implied that applications should be able to pass non-const pointers in. Of course, the working group’s resolution text doesn’t seem to even consider that the change breaks applications that supply const pointers instead, but there’s no way to allow both in a single prototype anyway. The resolution also explicitly says:

> For conformance to the current version of XSH4 either prototype is allowed.

Anyway, my point is: although the definition of `iconv` doesn’t explicitly say this, this history corroborates the common sense assumption that `iconv` never writes through this pointer and should accept a const pointer just fine.

Reply:
> older versions of the spec [did in fact include const](https://pubs.opengroup.org/onlinepubs/007908799/xsh/iconv.html)
> Of course, the working group’s resolution text doesn’t seem to even consider that the change broke applications that supply const pointers instead, but there’s no way to allow both in a single prototype anyway.

The version in the header is declared to be the normative form, thus `const char**` was to my understanding technically always incorrect (but the discrepancy quite confusing indeed). See [here](https://pubs.opengroup.org/onlinepubs/7908799/xsh/interfaces.html):

> will make all necessary symbols visible as described in the Headers section of this specification. 

I didn’t know both forms are allowed; though the current spec contains no mention of this. Are both allowed by the standard, or was this just a remark about their validation testsuite (for SUSv2)? I’m also not sure if an implementations is required to satisfy constraints of both variants (i.e. never modify input data) or if only those actually using `const` have to.

I’d love to be able to assume iconv never modifies the input memory, but there doesn't seem to be anything explicitly corroborating this. (Not does the description mention modifying inbuf data, which maybe ca be seen as implying its constness)

Reply:
> I didn’t know both forms are allowed; though the current spec contains no mention of this. Are both allowed by the standard, or was this just a remark about their validation testsuite (for SUSv2)?

My reading is that both forms aren’t allowed when claiming compliance with newer versions of the standard but both forms are (were) allowed if one wants to claim compliance with the older version(s) specifically.

> I’d love to be able to assume iconv never modifies the input memory, but there doesn't seem to be anything explicitly corroborating this. (Not does the description mention modifying inbuf data, which maybe ca be seen as implying its constness)

Yes, it’s frustrating that this isn’t actually specified, but I do think this is a risk I’m willing to take, personally, all things considered.

Reply:
By putting `const` into our API and leaving the iconv call an unmentioned implementation detail, we make this decision for all API users though and promise something which isn’t actually under out control, so I’m not sure we should.

By documenting users can decide whether to take their chances with `iconv` or not and by default (if not reading docs) the compiler will guide towards the guaranteed safe approach.
But if we constify this users might pass in read-only memory without ever taking a look at docs. If there’s some cursed iconv implementation actually modifying the input, this then turns into a direct path into UB.

Reply:
Given how truly cursed an input-modifying iconv would be, `const`ifying our API parameter _and_ noting potential iconv calls might also be ok, but `const`ifying without documenting iconv use...